ENH Add `HDBSCAN` as a new estimator in `sklearn.cluster`

Micky774 opened the pull request:
#### Reference Issues/PRs
Towards https://github.com/scikit-learn/scikit-learn/issues/24686

#### What does this implement/fix? Explain your changes.
Each change has been separately reviewed (see https://github.com/scikit-learn/scikit-learn/issues/24686 for details).

Edit: due to git shenanigans, some changes needed to be made within this PR. The novel changes included w/ this PR are:
1. Replaced `cnp.*_t` typing with `*_t` from `_typedefs.pxd`
2. Replaced `*.shape[0]` pattern with `len(*)` for `ndarray` objects
3. Trimmed unused variables (thanks to Cython linting pre-commit)

#### Any other comments?
cc: @thomasjpfan @jjerphan @glemaitre

Micky774 committed:
Include `HDBSCAN` as a sub-module for `sklearn.cluster` (#22616)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

Micky774 committed:
CLN Cleaned `cluster/_hdbscan/_linkage.pyx` (#24857)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
DOC Adds `HDBSCAN.dbscan_clustering` section to `plot_hdbscan.py` (#25538)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
ENH Extends outlier encoding scheme to `HDBSCAN.dbscan_clustering` (#24698)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
CLN Cleaned `cluster/_hdbscan/_reachability.pyx` (#24701)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

Micky774 committed:
CLN Update `cluster/_hdbscan/_tree.pyx` style and syntax (#25768)

Micky774 committed:
CLN Cleaned `TreeUnionFind` in `_hdbscan/_tree.pyx` (#25827)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
ENH `_hdbscan` `HIERARCHY` data type introduction (#25826)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
CLN BFS Style Improvement (#26096)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
CLN `_hdbscan/_tree.pyx` algorithms refactor (#26011)

Micky774 committed:
CLN HDBSCAN `_tree.pyx::do_labelling` refactor (#26101)

Micky774 committed:
Updated changelog

Micky774 committed:
Updated typing

jjerphan commented:
Since all PRs target the branch of this PR have been reviewed, merged, and are referenced in the issue, I thinking making the CI pass is sufficient to accept this PR.

Micky774 committed:
Updated typing, moved away from deprecated numpy API

thomasjpfan commented on sklearn/cluster/_hdbscan/_linkage.pyx:
```diff
@@ -110,23 +109,23 @@ cpdef cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst_from_data_matrix(
 
-    n_samples = raw_data.shape[0]
-    num_features = raw_data.shape[1]
+    n_samples = len(raw_data)
```
I think `.shape[0]` should still work. What is the reason for using `len` here?

Micky774 replied:
I ran into the following error, which I had (potentially incorrectly) attributed to the new numpy API:
```python-traceback

sklearn/cluster/_hdbscan/_tree.c: In function '__pyx_f_7sklearn_7cluster_8_hdbscan_5_tree_max_lambdas':
sklearn/cluster/_hdbscan/_tree.c:6453:39: error: 'PyArrayObject' {aka 'struct tagPyArrayObject'} has no member named 'dimensions'
 6453 |   __pyx_t_11 = (__pyx_v_condensed_tree->dimensions[0]);
      |                                       ^~
```

thomasjpfan replied:
For memoryviews, you can still use `.shape`. In this case, `raw_data` is a memoryview so it should still work.

For `cnp.ndarray`, you will have to avoid using `.shape`. I usually recommend going with `PyArray_SHAPE`.

Micky774 replied:
Fixed

Micky774 committed:
Updated shape semantics based on underlying data structure

Micky774 committed:
Doc fixes

Micky774 committed:
Updated shape semantics in remaining Cython code

Micky774 committed:
Merge branch 'main' into hdbscan_sync

Micky774 commented:
Should be good for review now @jjerphan @thomasjpfan ! For your convenience, here are the novel changes are included in these five commits: [changes](https://github.com/scikit-learn/scikit-learn/pull/26385/files/792485ab0e1ecfad93f9b42ebe228f84bc0fd0f9..87a43e55b68d0871fee0a4f015539da6dd077fcb)

These changes are comprised of:
1. Deletion of submodule-level `setup.py` files corresponding to old build strategy
2. Update to `setup.py` for building `_hdbscan` submodule
3. Sphinx doc appeasement in `clustering.rst`, `plot_hdbscan.py`, and `hdbscan.py`
4. Updated `sklearn/cluster/_hierarchical_fast.pyx::UnionFind` to avoid redeclaration of attributes and use new typings
5. Updated shape semantics in Cython files (`len` for lists/sets, `.shape[i]` for memoryviews, `PyArray_SHAPE` for `ndarray` objects)

jjerphan commented on sklearn/cluster/_hdbscan/_linkage.pyx:
```diff
@@ -0,0 +1,244 @@
+from ...utils._typedefs cimport intp_t, float64_t, int64_t, uint8_t
+
+cdef extern from "numpy/arrayobject.h":
+    ctypedef struct PyArrayObject
```
[`PyArrayObject` is listed in NumPy Cython API as of Cython 0.29.X](https://github.com/cython/cython/blob/8af954f812f78d5258cb8440b13cc6944eccf940/Cython/Includes/numpy/__init__.pxd#L237-L240) (but `PyArray_SHAPE` isn't).

Can `PyArrayObject` be `cimport`ed from `cnp`?

This suggestion applies to other files.

jjerphan commented on sklearn/cluster/_hdbscan/_tree.pxd:
```diff
@@ -0,0 +1,20 @@
+    float64_t value
+    intp_t cluster_size
+
+cdef extern from "numpy/arrayobject.h":
+    ctypedef struct PyArrayObject
+    intp_t * PyArray_SHAPE(PyArrayObject *)
```
The previous comment also applies here.

jjerphan commented on sklearn/cluster/_hierarchical_fast.pxd:
```diff
@@ -0,0 +1,9 @@
+    cdef intp_t[:] parent
+    cdef intp_t[:] size
+
+    cdef void union(self, intp_t m, intp_t n)
+    cdef intp_t fast_find(self, intp_t n)
```
Can we (technically) `noexcept`-qualify them?

```suggestion
    cdef void union(self, intp_t m, intp_t n) noexcept
    cdef intp_t fast_find(self, intp_t n) noexcept
```

jjerphan reviewed (approved):
LGTM.

Thank you for bringing HDBSCAN in scikit-learn, @Micky774!

Here are just minor comments.

Micky774 committed:
Use cimport for PyArrayObject and add noexcept for UnionFind

jjerphan commented on sklearn/cluster/_hdbscan/_linkage.pyx:
```diff
@@ -13,9 +13,9 @@ from ...cluster._hierarchical_fast cimport UnionFind
 from ...cluster._hdbscan._tree cimport HIERARCHY_t
 from ...cluster._hdbscan._tree import HIERARCHY_dtype
 from ...utils._typedefs cimport intp_t, float64_t, int64_t, uint8_t
+from numpy cimport PyArrayObject
```
Do you think occurrences of `PyArrayObject` be replaced by `cnp.PyArrayObject` instead?

Micky774 replied:
I prefer import types directly (e.g. `HIERARCHY_t`) but I don't have a huge preference. I have updated them now :)

jjerphan replied:
So do I, I think that we tend to import NumPy's symbol using `cnp.` everytime (hence my question). But that's minor as a comment.

Micky774 committed:
Updated cimport strategy

jjerphan commented:
I think it is reasonable to wait for a third reviewer to approve this PR before merging it.

Micky774 commented:
cc: @scikit-learn/core-devs in case anyone would like to donate a third review :)

adrinjalali commented on sklearn/cluster/_hdbscan/hdbscan.py:
```diff
@@ -0,0 +1,788 @@
+
+
+class HDBSCAN(ClusterMixin, BaseEstimator):
+    """Cluster data using hierarchical density-based clustering.
```
would also be nice to have a link to the example where DBSCAN and HDBSCAN are compared here, instead of just relying on the auto detection of all examples, which doesn't make examples really discoverable.

Micky774 replied:
Done!

adrinjalali reviewed (commented):
The copyright here: https://github.com/scikit-learn-contrib/hdbscan/blob/master/LICENSE isn't really BSD, and doesn't allow to just use it, to me it seems we need a copy of that file here? Unless that copyright is changed or we somehow get a written permission from all contributors to that package.

I left a few comments just to make a few things more readable, otherwise I'm happy to merge.

Micky774 committed:
Added copy of license

Micky774 commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -79,6 +79,9 @@
     "min_samples": 7,
     "xi": 0.05,
     "min_cluster_size": 0.1,
+    "allow_single_cluster": True,
+    "hdbscan_min_cluster_size": 15,
+    "hdbscan_min_samples": 3,
```
Using separate keys would make the dataset-specific parameter updates a bit more verbose and potentially decrease clarity, so perhaps enforcing namespace via the string itself (e.g. `hdbscan_min_samples`) for the other params where possible would be safer.

With that being said, I suspect it would be better to relegate that to a separate PR to keep this one ~~simple~~ somewhat manageable :)

adrinjalali replied:
Could you please then open an issue for this so that we don't forget?

Micky774 committed:
Updated documentation of helper functions/methods

Micky774 committed:
Improved testing documentation

Micky774 commented on sklearn/cluster/_hdbscan/hdbscan.py:
```diff
@@ -0,0 +1,788 @@
+    if issparse(distance_matrix) and distance_matrix.format != "csr":
+        # we need CSR format to avoid a conversion in `_brute_mst` when calling
+        # `csgraph.connected_components`
+        distance_matrix = distance_matrix.tocsr()
```
Odd, this should be hit with `test_hdbscan_sparse_distance_matrix` when `sparse_constructor=sparse.csc_matrix` I believe.

Micky774 committed:
Updated documentation to explicitly include example

adrinjalali reviewed (approved):
PRETTY KEWL!!!
